graph/path/path.go

```diff
@@ -0,0 +1,296 @@
+// Copyright 2014 The Cayley Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package path
+
+import (
+	"github.com/google/cayley/graph"
+	"github.com/google/cayley/graph/iterator"
+	"github.com/google/cayley/quad"
+)
+
+type morphism struct {
+	Name     string
+	Reversal func() morphism
+	Apply    graph.ApplyMorphismFunc
+}
+
+type Path struct {
```

Comment:
The only panics are programmer error -- passing wrong types, making invalid assumptions, etc. With a richer type system, this would be in the type checker, but panic is my solution to the madness.